src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -442,31 +525,186 @@ def restrict(self, index: common.AnyIndexSpec) -> NdArrayConnectivityField:
+            )
+        conn_map[conn.codomain] = conn
+
+        dim_idx = data.domain.dim_index(conn.codomain)
+        current_range: common.UnitRange = data.domain.ranges[dim_idx]
+        new_conn_ranges = connectivity.inverse_image(current_range).ranges
+        new_ranges = tuple(r & s for r, s in zip(new_ranges, new_conn_ranges))
+
+    conns_dims = [c.domain.dims for c in conn_map.values()]
+    for i in range(len(conns_dims) - 1):
+        if conns_dims[i] != conns_dims[i + 1]:
+            raise ValueError(
+                f"All remapping connectivities must have the same dimensions, got: {conns_dims}."
+            )
+
+    new_domain = common.Domain(dims=data.domain.dims, ranges=new_ranges)
+
+    # Create identity connectivities for the missing domain dimensions
+    for dim in data.domain.dims:
+        if dim not in conn_map:
+            conn_map[dim] = utils.first(
+                _identity_connectivities(new_domain, [dim], cls=type(connectivity))
+            )
+
+    # Take data
+    take_indices = tuple(conn_map[dim].ndarray for dim in data.domain.dims)
+    new_buffer = data._ndarray.__getitem__(take_indices)
+
+    return data.__class__.from_array(
+        new_buffer,
+        domain=new_domain,
+        dtype=data.dtype,
+    )
+
+
+def _remapping_premap(data: NdArrayField, connectivity: common.ConnectivityField) -> NdArrayField:
+    new_dims = {*connectivity.domain.dims} - {connectivity.codomain}
+    if repeated_dims := (new_dims & {*data.domain.dims}):
+        raise ValueError(f"Remapped field will contain repeated dimensions '{repeated_dims}'.")
+
+    # Compute the new domain
+    dim = connectivity.codomain
+    dim_idx = data.domain.dim_index(dim)
+    if dim_idx is None:
+        raise ValueError(f"Incompatible index field, expected a field with dimension '{dim}'.")
+
+    current_range: common.UnitRange = data.domain[dim_idx][1]
+    new_ranges = connectivity.inverse_image(current_range)
+    new_domain = data.domain.replace(dim_idx, *new_ranges)
+
+    # Perform premap:
+    xp = data.array_ns
+
+    # 1- first restrict the connectivity to the new domain
+    restricted_connectivity_domain = common.Domain(*new_ranges)
+    restricted_connectivity = (
+        connectivity.restrict(restricted_connectivity_domain)
+        if restricted_connectivity_domain != connectivity.domain
+        else connectivity
     )
+    assert isinstance(restricted_connectivity, common.ConnectivityField)
 
+    # 2- then compute the index array
+    new_idx_array = xp.asarray(restricted_connectivity.ndarray) - current_range.start
 
-def _hypercube(
+    # 3- finally, take the new array
+    new_buffer = xp.take(data._ndarray, new_idx_array, axis=dim_idx)
+
+    return data.__class__.from_array(
+        new_buffer,
+        domain=new_domain,
+        dtype=data.dtype,
+    )
+
+
+_ConnT = TypeVar("_ConnT", bound=common.ConnectivityField)
+
+
+def _identity_connectivities(
```

Comment:
it seems rewriting to `_identity_connectivity()` (singular) could make sense, then the call site is just a list comprehension instead of a for-loop here

Reply:
looking at the call-site, it seems the current state is a result of some refactoring because it doesn't make sense now

Reply:
Done